### ISSUE TITLE
Batch node expects the time to be in nanoseconds

### ISSUE THREAD

**Issue by colinl** [mblackstock/node-red-contrib-influxdb#44]
When writing to influx using the batch node and providing a timestamp the readme says that it should be in milliseconds or should be a javascript Date object.  It appears that in fact the time has to be provided in nanoseconds.

**Comment by malcolmwhat** [comment]
This comes down to a difference in the wording of the documentation, and is a minor bug in this repositories code (or a bug in the readme depending on how you look at it). The default behavior of the underlying package is to use `millisecond` time precision **if you specify a timestamp in your payload**. I personally read this as the default being milliseconds in general, but their code stands up to this nuanced documentation, so the problem is in this repo and not theirs.. Though it might be worth submitting a PR to get them to clarify or bolden their documentation when they are going to word things that way..

From the README of this repo:
> By default the node will write timestamps using ms precision since that's what JavaScript gives us. if you specify the timestamp as a Date object, we'll convert it to milliseconds. If you provide a string or number as the timestamp, we'll pass it straight into Influx to parse.

From the [writePoints documenation in node-influx](https://node-influx.github.io/class/src/index.js~InfluxDB.html) (emphasis is mine, not theirs)
> A note when using **manually-specified times and precisions**: by default we write using the ms precision since that's what JavaScript gives us. You can adjust this. **However, there is some special behavior if you manually specify a timestamp in your points**:
>
>    if you specify the timestamp as a Date object, we'll convert it to milliseconds and manipulate it as needed to get the right precision
    if provide a INanoDate as returned from toNanoTime or the results from an Influx query, we'll be able to pull the precise nanosecond timestamp and manipulate it to get the right precision

There is some testing information below if you want to follow my dig into this :)
.
.
.


When passing this down to that exact documented endpoint 
```
[{"fields":{"ran":0.399385292240906},"tags":{"tag":"Test1"},"measurement":"kpi"},{"fields":{"ran":0.026788376114670998},"tags":{"tag":"Test2"},"measurement":"kpi"}]
```

we get nanosecond precision, even if we do not include any timestamp or precision in the payloads, for example:

[{"fields":{"ran":0.01789827873471017},"tags":{"tag":"Test1"},"measurement":"kpi"},{"fields":{"ran":0.319917888149849},"tags":{"tag":"Test2"},"measurement":"kpi"}]
{}

Resulted in [httpd] 127.0.0.1 - root [14/Sep/2019:12:30:06 -0400] "POST /write?db=testmeasurement&p=%5BREDACTED%5D&precision=n&rp=&u=root HTTP/1.1" 204 0 "-" "-" e96fc302-d70c-11e9-803f-98fa9b115977 10960

### More details if you want to have a look



Node Red Config:
```
[
    {
        "id": "824509dd.6ca09",
        "type": "inject",
        "z": "38702606.9067ca",
        "name": "",
        "topic": "",
        "payload": "",
        "payloadType": "date",
        "repeat": "",
        "crontab": "",
        "once": false,
        "onceDelay": 0.1,
        "x": 120,
        "y": 280,
        "wires": [
            [
                "7879c3fa.c58984"
            ]
        ]
    },
    {
        "id": "29c4119e.05bb1e",
        "type": "influxdb batch",
        "z": "38702606.9067ca",
        "influxdb": "fa665d.ebb9a1a",
        "precision": "",
        "retentionPolicy": "",
        "name": "Write msgs",
        "x": 490,
        "y": 280,
        "wires": []
    },
    {
        "id": "7879c3fa.c58984",
        "type": "function",
        "z": "38702606.9067ca",
        "name": "Batch Write",
        "func": "// Time in milliseconds.\nlet x = (new Date()).getTime();\n\nnode.error(x);\n\nmsg.payload = [{\n    \"timestamp\": x,\n    \"fields\":{\n        \"ran\": Math.random()\n    },\n    \"tags\":{\n        \"tag\": \"Test1\"\n    },\n    \"measurement\": \"kpi\"\n},\n{\n    \"timestamp\": x,\n    \"fields\":{\n        \"ran\": Math.random()\n    },\n    \"tags\":{\n        \"tag\": \"Test2\"\n    },\n    \"measurement\": \"kpi\"\n}];\n\nreturn msg;",
        "outputs": 1,
        "noerr": 0,
        "x": 290,
        "y": 280,
        "wires": [
            [
                "29c4119e.05bb1e"
            ]
        ]
    },
    {
        "id": "fa665d.ebb9a1a",
        "type": "influxdb",
        "z": "",
        "hostname": "127.0.0.1",
        "port": "8086",
        "protocol": "http",
        "database": "testmeasurement",
        "name": "",
        "usetls": false,
        "tls": ""
    }
]
```

`time` logged from flow and used in batch write when on default: 1568476243928

Resulting influxd log:
```
[httpd] 127.0.0.1 - root [14/Sep/2019:11:50:43 -0400] "POST /write?db=testmeasurement&p=%5BREDACTED%5D&precision=n&rp=&u=root HTTP/1.1" 204 0 "-" "-" 691641ec-d707-11e9-801e-98fa9b115977 5981
```

Clearly using nanosecond precision.  Here is the result for the query on that measurement.
```
name: kpi
time          ran                tag
----          ---                ---
1568476243928 0.6908260375206823 Test2
1568476243928 0.5239582943471468 Test1
```

**Comment by mblackstock** [comment]
Will setting precision to 'ms' when it is not specified explicitly break any existing flows?

**Comment by malcolmwhat** [comment]
Use cases:

| Setting for Precision        | Type of timestamp           | Change?  |
| ------------- |:-------------:| -----:|
| Explicitly set      | Don't care | None |
| Default      | Date object |   No change (was always making ms precision) |
| Default      | (Dateobj).getTime() | This would be broken previously and is the crux of this issue (I think). Would now result in correct translation from the getTime() -> UTC time in db |
| Default      | None specified | This would now write with ms precision instead of ns precision. Wasn't broken previously but we would have reduced default precision. |

I can't see anyone having got very far with the third option so I wouldn't consider that a break in the behavior, however if we consider the last option to be no good then the better resolution is to clarify the README.

**Comment by mblackstock** [comment]
Sorry to take so long to reply!

If we default to ms when the precision is not specified, this will be different behaviour from what the influxdb docs indicate will happen (i.e. it uses nanosecond precision).  Might not be that bad, but it makes me slightly uncomfortable since we want things to be consistent with their docs where possible.

Folks should know that date.getTime() is a number in milliseconds, and if they don't specify the precision, given the default is nanoseconds, there will be a problem.  I think should improve the docs there to make this explicit.

I suggest a README change to say something like:

If you provide a string or number as the timestamp, we'll pass it straight into Influx to parse using the specified precision or the default precision in nanoseconds if it is left unspecified.

>>Note that the default precision is nanoseconds, so if you pass in a number such as date.getTime(), and do not specify millisecond precision, your timestamp will be orders of magnitude incorrect.

**Comment by malcolmwhat** [comment]
This can be closed as can #47
